dask/array/core.py

```diff
@@ -889,65 +889,61 @@ def store(sources, targets, lock=True, regions=None, compute=True,
     sources_dsk = sharedict.merge(*[e.__dask_graph__() for e in sources])
     sources_dsk = Array.__dask_optimize__(
         sources_dsk,
-        [e.__dask_keys__() for e in sources]
+        list(core.flatten([e.__dask_keys__() for e in sources]))
     )
 
-    tgt_dsks = []
-    store_keys = []
-    store_dsks = []
-    if return_stored:
-        load_names = []
-        load_dsks = []
-    for tgt, src, reg in zip(targets, sources, regions):
-        # if out is a delayed object update dictionary accordingly
+    # Optimize all targets together
+    targets_keys = []
+    targets_dsk = []
+    for e in targets:
         try:
-            each_tgt_dsk = {}
-            each_tgt_dsk.update(tgt.dask)
-            tgt = tgt.key
+            targets_keys.extend(e.__dask_keys__())
+            targets_dsk.append(e.__dask_graph__())
         except AttributeError:
-            each_tgt_dsk = {}
+            pass
+
+    targets_dsk = sharedict.merge(*targets_dsk)
+    targets_dsk = Delayed.__dask_optimize__(targets_dsk, targets_keys)
 
+    load_stored = (return_stored and not compute)
+
+    store_keys = []
+    store_dsk = []
+    for tgt, src, reg in zip(targets, sources, regions):
+        tgt = getattr(tgt, "key", tgt)
```

Comment:
Can you explain what this does? It's not clear to me what case works with a target with a `key` attribute. If this is for supporting `Delayed` targets, I'd prefer an explicit instance check rather than an attribute check as:
- It's more apparent what the intent is
- Won't fail for targets that just happen to have a `key` attribute.

Reply:
Right, as you guessed this is dealing with `Delayed` targets. The original code actually used `try...except AttributeError:...`. Was originally introduced in PR ( https://github.com/dask/dask/pull/2181 ). So thought this was some small improvement. Though agree we should be verifying things actually are `Delayed`. That said, we might want this check a bit earlier as there is a call to `__dask_optimize__` for these objects.

Reply:
Makes sense to me. Perhaps something like this earlier:

```python
for t in targets:
    if isinstance(t, Delayed):
        targets2.append(t.key)
    elif is_dask_collection(t):
        raise ValueError("Targets must be either Delayed objects or array-likes")
    else:
        targets2.append(t)
```

Then use `targets2` unconditionally in this loop.

Reply:
Should also probably update the docstring for `store` to add that `targets` can be instances of `Delayed`.